### DISCUSSION TITLE
Explain it to me like I'm 35...

### DISCUSSION THREAD

**Discussion by PLK2** [rust-bitcoin/rust-bitcoin#1036] [Q&A] [answered]
I'm a longtime bitcoiner but greenhorn programmer. Can someone help me better understand the initiative to rewrite Bitcoin in Rust?
1. What are the general goals and benefits?
2. If Rust is so much better and safer than C++ (or so I hear), why hasn't this already been done? Especially with $700b in bitcoin on the line.
3. What are the general benefits of having bitcoin written in various languages? C++, Go, Rust, etc? Doesn't that divide resources to maintain them all? Or is it helping decentralize the code in some way?

**Comment by TheBlueMatt** [comment, marked as answer]
I don't think there's a desire to "rewrite Bitcoin (Core) in rust", but there's definitely a desire to use safe, efficient languages (like rust!) for Bitcoin, lightning, etc wallets and other software.

As for Bitcoin core, the issue is always consensus - every node agree exactly on if each transaction is valid or invalid, including all strange edge cases on transactions not yet seen. It's much, much harder than it looks, and rewriting the core consensus rules in another language is likely not practical without some rather massive effort in formal analysis. That said, for the non-consensus parts, it'd be nice to replace them or offer alternatives, see the libbitcoinkernel project.

**Reply by PLK2** [reply]
Oh I see. I think I misunderstood, thinking Core was transitioning over to Rust.

That makes sense why it would be such a massive undertaking to rewrite the consensus rules, given how slowly/carefully they've evolved in C++ with all their interdependencies. I imagine it's like TCP/IP--could be better, but it is what it is at this point.

So is the goal for the rust-bitcoin repository to provide a library of Rust code for bitcoin-related projects to use? And interface with the bitcoin network, via Rust?

Thanks for replying!

**Reply by TheBlueMatt** [reply]
I dunno, rust is also a relatively young language and a bit less well-defined. For consensus rules, you really may want ada over rust, but C/C++ is a good second - the definition of it is still quite ridged.

In any case, yes, the intent of this repo is not to implement consensus rules, but to provide primitives which can be reused to implement Bitcoin software - it is used by several downstream projects to power Bitcoin and lightning wallets, blockchain indexers and block explorers, network scanners, etc.

**Reply by Kixunil** [reply]
I marked this as the answer, feel free to change if you disagree.

**Comment by dpc** [comment]
> Can someone help me better understand the initiative to rewrite Bitcoin in Rust?

Right now there's no immediate initiative to replace Bitcoin Core with anything written in 100% Rust.

I think it will gradually and eventually happen in the long run (I'd say 20 years), though at that point it might not even be Rust. Maybe it will be some Rust++ or Rust2, or whatever else things will evolve into. And it might be a hybrid solution of some kind, and so on.

> What are the general benefits of having bitcoin written in various languages? C++, Go, Rust, etc? Doesn't that divide resources to maintain them all? Or is it helping decentralize the code in some way?

It does divide resources, and that's a big drawback. It also makes consensus changes (soft-forks) potentially harder (which is both a drawback and benefit, I guess), as multiple projects would have to implement them.

The benefits I see is:

* minimizing possibility of a catastrophic network failure due to software bug; such bugs are rare, but can happen (see [CVE-2018-17144](https://bitcoincore.org/en/2018/09/20/notice/)), by making the network more heterogeneous, 
* increases the mindshare be allowing more people to participate,
* makes building some thing on Bitcoin easier, by being able to customize the node itself.

> If Rust is so much better and safer than C++ (or so I hear), why hasn't this already been done?

Rust is relatively novel thing. Also, by no means it is magic wand, making everything better. Such things take a lot of time, and also require certain amount of social buy-in: there has to be enough people convinced and willing to work towards such goal.

**Reply by PLK2** [reply]
Thank you for helping me make sense of all this 🙏
